### PR TITLE
Add cloudflare guide on DOCS.md

### PR DESCRIPTION
**Pull request by Black2vs2**
Following this discussion https://github.com/llamafilm/tesla-http-proxy-addon/discussions/23 i got the addon finally working.

With the contribution of @baylanger and @SlowCar999 i made the guide for cloudflare setup.

### CODE REVIEW

**Comment by Black2vs2** [comment]
@llamafilm Any way this can be merged into your repo?

File: tesla_http_proxy/DOCS.md [hunk 1]
```diff
@@ -14,38 +14,103 @@ servers: nginx_proxy/*.conf
...
+Setting up Tesla HTTP Proxy through cloudflare involves the following steps (all the steps are made with the available addons in home assistant, nothing external. Also it avoids the necessity to open ports on the router):
+
+1. Install **Cloudflared** addon. Then configure like this and start the addon.
+
+```yml
+external_hostname: ha.example.com
+additional_hosts:
+  - hostname: tesla.example.com
+    service: https://homeassistant-internal.example.com
+```
```
Comment (baylanger) [inline tesla_http_proxy/DOCS.md:58]: Does the add-on use the CF API to create the tunnel OR does the end user needs to configure that as well?

File: tesla_http_proxy/DOCS.md [hunk 2]
```diff
@@ -14,38 +14,103 @@ servers: nginx_proxy/*.conf
...
+
+2. Check on the **Cloudflare dashboard** that the two CNAME DNS Records appeared. If not read the **cloudflared** logs for errors.
```
Comment (baylanger) [inline tesla_http_proxy/DOCS.md:60]: Are those records automatically created by the add-on?

File: tesla_http_proxy/DOCS.md [hunk 3]
```diff
@@ -14,38 +14,103 @@ servers: nginx_proxy/*.conf
...
+   ![image](https://github.com/llamafilm/tesla-http-proxy-addon/assets/1372028/42e637a8-4428-41c7-b244-099aa6844216)
+3. Create an `A` type record with name `homeassistant-internal` and IPv4 address pointing to the `LOCAL IP ADDRESS` of the machine hosting the homeassistant instance. Disable `proxied`.
+4. Install **Let's Encrypt** addon. (can be done with other addons, check discussions on the repo for alternatives). Then configure like this and make sure to configure the `cloudflare_api_token`. Make sure the `keyfile` and `certfile` aren't already been created. If so, delete them and proceed starting the addon. Read the logs to see when the files are created (takes more than 60 seconds).
```
Comment (baylanger) [inline tesla_http_proxy/DOCS.md:63]: Might be worth adding the URL to the [Discussions](https://github.com/llamafilm/tesla-http-proxy-addon/discussions) for those who aren't quite github familiar.

File: tesla_http_proxy/DOCS.md [hunk 4]
```diff
@@ -14,38 +14,103 @@ servers: nginx_proxy/*.conf
...
+
+```yml
+domains:
+  - ha.example.com
+  - "*.ha.example.com"
+  - tesla.example.com
+  - homeassistant-internal.example.com
+email: null@null.com
```
Comment (baylanger) [inline tesla_http_proxy/DOCS.md:71]: To make it more clear I would use something like `your-email@domain.com`

File: tesla_http_proxy/DOCS.md [hunk 5]
```diff
@@ -14,38 +14,103 @@ servers: nginx_proxy/*.conf
...
+
+2. Check on the **Cloudflare dashboard** that the two CNAME DNS Records appeared. If not read the **cloudflared** logs for errors.
+   ![image](https://github.com/llamafilm/tesla-http-proxy-addon/assets/1372028/42e637a8-4428-41c7-b244-099aa6844216)
+3. Create an `A` type record with name `homeassistant-internal` and IPv4 address pointing to the `LOCAL IP ADDRESS` of the machine hosting the homeassistant instance. Disable `proxied`.
+4. Install **Let's Encrypt** addon. (can be done with other addons, check discussions on the repo for alternatives). Then configure like this and make sure to configure the `cloudflare_api_token`. Make sure the `keyfile` and `certfile` aren't already been created. If so, delete them and proceed starting the addon. Read the logs to see when the files are created (takes more than 60 seconds).
+
+```yml
+domains:
+  - ha.example.com
+  - "*.ha.example.com"
```
Comment (baylanger) [inline tesla_http_proxy/DOCS.md:68]: For some people this might be confusing... for example, Nabu Casa afaik people don't need those 2 entries above.

File: tesla_http_proxy/DOCS.md (hunk 1 above)
Comment (Black2vs2) [reply tesla_http_proxy/DOCS.md:58]: The addon uses CF API, no other input needed to create the tunnel

File: tesla_http_proxy/DOCS.md (hunk 2 above)
Comment (Black2vs2) [reply tesla_http_proxy/DOCS.md:60]: Yes

File: tesla_http_proxy/DOCS.md [hunk 6]
```diff
@@ -14,38 +14,103 @@ servers: nginx_proxy/*.conf
...
+keyfile: privkey.pem
+cloudflare: true
+customize:
+  active: true
+  default: nginx_proxy_default*.conf
+  servers: nginx_proxy/*.conf
+```
+
+5. Install **Tesla HTTP Proxy** addon. Configure it as below and then follow all the steps in the `DOCS` making sure to fill the `client_id` and `client_secret`.
+
```
Comment (baylanger) [inline tesla_http_proxy/DOCS.md:95]: Might be worth to mention the user to pick his `region`.

File: tesla_http_proxy/DOCS.md [hunk 7]
```diff
@@ -54,10 +119,23 @@ nxjvrm2M6uKfDEYS52ITVVbzqGMzzbKCO/tuu78432jU6Z96BNR8NSoRXg==
 -----END PUBLIC KEY-----
 ```
 
-You should have a config file at `/share/nginx_proxy/nginx_tesla.conf` that hosts the static public key file.  You may need to modify this file depending on your SSL config.  
+- You should have a config file at `/share/nginx_proxy/nginx_tesla.conf` that hosts the static public key file. You may need to modify this file depending on your SSL config.
```
Comment (baylanger) [inline tesla_http_proxy/DOCS.md:122]: The config file doesn't host the static pub key file but has a setting to specify the path to the pub/priv key. Unless I'm wrong this probably need to be rephrased.

File: tesla_http_proxy/DOCS.md [hunk 8]
```diff
@@ -54,10 +119,23 @@ nxjvrm2M6uKfDEYS52ITVVbzqGMzzbKCO/tuu78432jU6Z96BNR8NSoRXg==
...
+
+- If you get `login_required` error when trying to send API commands, it's likely because you tried to reuse the refresh token more than once. Try fully removing the Tesla integration from HA and adding it back again.
+
+- When you enroll the public key in the vehicle, if you don't get a prompt, try moving within BLE range of the vehicle.
```
Comment (baylanger) [inline tesla_http_proxy/DOCS.md:126]: I didn't think BLE connection to the car was a requirement but I could be wrong... perhaps this was mentioned somewhere in a Discussion?!

**Comment by baylanger** [comment]
There's a whole lot of unexpected things happening in my life since many weeks. I had and still need time for myself and family but decided to take a few minutes to read / comment the PR.

@Black2vs2 Thanx for putting this together.

File: tesla_http_proxy/DOCS.md (hunk 8 above)
Comment (luca-sartori-zupit) [reply tesla_http_proxy/DOCS.md:126]: This comment is done by the author of the repo, not me.
It probably needs to be done in order to wake the car

File: tesla_http_proxy/DOCS.md (hunk 7 above)
Comment (luca-sartori-zupit) [reply tesla_http_proxy/DOCS.md:122]: Fixed, it's a author's suggestion

File: tesla_http_proxy/DOCS.md [hunk 9]
```diff
@@ -14,38 +14,103 @@ servers: nginx_proxy/*.conf
...
+6. Check on the **Cloudflare dashboard** that in the `SSL/TLS Edge Certificates` there's an active universal certificate for both `example.com` and `*.example.com`. Make sure that in the `SSL/TLS Overview` page the encryption mode is set to `Flexible`
+
 ## Troubleshooting
 
-Check the add-on logs to see what's happening.
+- Check the add-on logs to see what's happening.
 
-From the add-on Web UI there is a link to test your public key HTTPS endpoint.  On iOS this will cause a dialog about "trying to download a configuration profile" because it incorrectly identifies the public key as such.  On a desktop browser it should display the contents of your public key, similar to this:
+- From the add-on Web UI there is a link to test your public key HTTPS endpoint. On iOS this will cause a dialog about "trying to download a configuration profile" because it incorrectly identifies the public key as such.
+  If the certificates aren't showing up (most importantly because of some errors) make sure to clear the cache of the browser or just wait some hours.
```
[the remark this answers is not in the document]
Comment (luca-sartori-zupit) [reply tesla_http_proxy/DOCS.md:112]: I don't think so, author made a really easy to use button in the addon interface in home assistant.
Wait some hours for the cache of cloudflare to be gone

File: tesla_http_proxy/DOCS.md (hunk 5 above)
Comment (luca-sartori-zupit) [reply tesla_http_proxy/DOCS.md:68]: True. Tho using them shouldn't be a problem. This specific guide is for cloudflare users tho

**Comment by Black2vs2** [comment]
Thanks for the contribution @baylanger, done some fixes and answered some comments. I hope you get better and proceed with your life! Take your time :)

**Comment by llamafilm** [comment]
Thank you for the contribution.  I'm going to sign up for a cloudflare account and follow these instructions to see how it goes before approving.

File: tesla_http_proxy/DOCS.md (hunk 8 above)
Comment (llamafilm) [reply tesla_http_proxy/DOCS.md:126]: I don't know why BLE is relevant, but it was mentioned by an Android user here: https://github.com/llamafilm/tesla-http-proxy-addon/discussions/58#discussioncomment-8889866

File: tesla_http_proxy/DOCS.md (hunk 4 above)
Comment (llamafilm) [reply tesla_http_proxy/DOCS.md]: If email is not required then `null` is a good indicator of that.  Is it possible to omit that field instead?

File: tesla_http_proxy/DOCS.md (hunk 6 above)
Comment (llamafilm) [reply tesla_http_proxy/DOCS.md:95]: I suggest removing this section, since most people will configure using the GUI, and instructions are already written above.

File: tesla_http_proxy/DOCS.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (llamafilm) [reply tesla_http_proxy/DOCS.md:112]: Maybe reword: "or just wait a few hours for CloudFlare's cache to update."

**Comment by baylanger** [comment]
I think I forgot to mention, I don't use the add-on but an external CF Docker instance, that's the reason I raised some questions. After learning how simple the add-on works, I might end up using it and dropping the "external" Docker instance.

File: tesla_http_proxy/DOCS.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (baylanger) [reply tesla_http_proxy/DOCS.md:112]: Oh ... it is Clouflare caching system that is involved in here? I don't have time to go in my account and check CF's "Cache engine" options, etc...

Would adding a Cache-Control ~ 5 minutes header in the HTTP request work or it's better to leave it as-is?

**Comment by llamafilm** [comment]
@Black2vs2 thanks for writing this guide.  This is my first time using CloudFlare and I like it.  I'm considering switching my domain over there, to leverage its caching and DDoS protection, but I would not be able to use "proxied" mode because I expose other nonstandard ports on my firewall.  I've also read some concerns above privacy and MITM, although I think that only applies to "Flexible" SSL mode, which sounds like a terrible idea overall.  I need to do some more reading.

In current form, I think this guide belongs better as a Wiki page.  Over time, I hope more users write wiki pages to explain their unique setups.  But I'm trying to decide if it would make sense to make this the default recommendation, and make CloudFlare a dependency of this addon, just like Nginx is.  I want to reduce the barrier to entry for onboarding new users.  What do you think?

Either way, I suggest renaming the guide to CloudFlare Argo Tunnel, since CloudFlare as a company offers many different tools and services.

**Comment by baylanger** [comment]
I agree a WiKi makes sense, I believe it can easily be edited by anyone.

Some people might not like being forced to use CF but ... considering creating a "free" CF account is easy, the CF add-on makes everything simple to setup and that probably a lot of T owners running HA are hoping for a "simple way" to get everything up and running easily.

Also, this CF Argo Tunnel likely doesn't interface at all with a Nabu Casa setup.

**Comment by Black2vs2** [comment]
Good idea to move it to a wiki page, tho i would like to add the link to the wiki in the DOCS.md to keep it as visible as the rest of the docs.
Cloudflared is really the simplest way to get it to work (didn't try the others, but as long as it involves using command line to run a docker, it is a step that not all users are ready/able to take)

You can still access every port you need. You just need to have a reverse proxy for each port you need (services in cloudflared) am i wrong?
> but I would not be able to use "proxied" mode because I expose other nonstandard ports on my firewall

**Comment by Black2vs2** [comment]
I'm not familiar with creating/using wiki, i cannot do one on this repo. Is there a pull request i can do for that or else?

**Comment by baylanger** [comment]
> I'm not familiar with creating/using wiki on GH, i cannot create one on this repo. Is there a pull request i can do for that or else?
>
The repo might have too much restriction. Either it's wide open (anyone can create/edit) and it's probably possible to add individual accounts. @llamafilm can add you and perhaps myself? I'm still around since yesterday trying to help 1 person... as I mentioned, I'll step away "soon" for the coming weeks even if I enjoy helping this project.

**Comment by baylanger** [comment]
Currently it's probably set to only allow Collaborators to create/edit WiKis.



[alandtse/tesla](https://github.com/alandtse/tesla/wiki) repo allows anyone. If it's wide open, might be a good idea to have mention somewhere (README or DOCS or a WiKi?) not to create a WiKi w/o consensus. Not sure how @llamafilm wants to handle it.

File: tesla_http_proxy/DOCS.md [hunk 10]
```diff
@@ -14,38 +14,103 @@ servers: nginx_proxy/*.conf
...
 > [!NOTE]
 > Your Tesla app must already be key-paired with the car.
 
 Configure the [Tesla integration](https://github.com/alandtse/tesla) to use this proxy. It should pre-fill the Client ID, URL, and certificate for you by reading them from this addon.
 
+## Specific setup guides
+
+### Cloudflare
+
+Setting up Tesla HTTP Proxy through cloudflare involves the following steps (all the steps are made with the available addons in home assistant, nothing external. Also it avoids the necessity to open ports on the router):
```
Comment (baylanger) [inline tesla_http_proxy/DOCS.md:49]: should be Cloudflare w/ C* (sorry I'm picky :p )

File: tesla_http_proxy/DOCS.md [hunk 11]
```diff
@@ -14,38 +14,103 @@ servers: nginx_proxy/*.conf
...
+
+1. Install **Cloudflared** addon. Then configure like this and start the addon.
```
Comment (baylanger) [inline tesla_http_proxy/DOCS.md:51]: Just to make sure, are you referring to the [Cloudflare add-on](https://www.home-assistant.io/integrations/cloudflare/) or the [Cloudflare Integration](https://www.home-assistant.io/integrations/cloudflare/) ?

Consider adding the link to either the integration or the add-on. Though it's possible either works?

Comment (baylanger) [reply tesla_http_proxy/DOCS.md:51]: @llamafilm yesterday, when you did your setup with CF... you used the integration or add-on?

**Comment by llamafilm** [comment]
I just opened access so you should be able to make a page now.
I added a link to Wiki from the bottom of docs. Maybe you can make a link from the FAQ to the new page about Cloudflare?

True, I could make a tunnel service for each other port/service. But then those other services become dependent on Home Assistant. I want the other services to remain up even if HA crashes. Plus the privacy issues, Cloudflare can MITM all your https traffic, so I will stick with traditional port forwarding.

**Comment by Black2vs2** [comment]
Created the wiki page. We can close the PR i think

File: tesla_http_proxy/DOCS.md (hunk 11 above)
Comment (Black2vs2) [reply tesla_http_proxy/DOCS.md:51]: It is actually very clear, Install Cloudflare**D** (the integration is cloudflare).
Also the word after cloudflared is "addon" so it's very clear ahaha

**Comment by baylanger** [comment]
Are we all on the same page that for Cloudflare service, the end user must own a private domain? If so might be a good idea to make it clear.

**Comment by Black2vs2** [comment]
@llamafilm Any specific reason why the wiki page isn't linked to the readme of the project?
I didn't think about even checking if there were any pages when i first had some problems because i tought all the useful links should be on the primary readme.
Can we add it in this PR and close it?

**Comment by llamafilm** [comment]
It's linked at the bottom of DOCS.md but if you want to add it to the wiki in this PR that's fine

**Comment by llamafilm** [comment]
It's linked at the bottom of DOCS.md but if you want to add it to the readme in this PR that's fine
